vmm/src/memory_manager.rs

```diff
@@ -1308,30 +1347,80 @@ impl MemoryManager {
+            let mut range: Option<(Range, bool)> = None;
+            if let Some(virtio_mem_zone) = memory_zone.virtio_mem_zone_mut() {
+                let virtio_mem_region = virtio_mem_zone.region();
+                match size.cmp(&virtio_mem_zone.hotplugged_size) {
+                    Ordering::Greater => {
+                        range = Some((
+                            Range {
+                                iova: virtio_mem_region.start_addr().raw_value()
+                                    + virtio_mem_zone.hotplugged_size,
+                                size: size - virtio_mem_zone.hotplugged_size,
+                                user_addr: virtio_mem_region.as_ptr() as u64
+                                    + virtio_mem_zone.hotplugged_size,
+                            },
+                            true,
+                        ));
+                    }
+                    Ordering::Less => {
+                        range = Some((
+                            Range {
+                                iova: virtio_mem_region.start_addr().raw_value() + size,
+                                size: virtio_mem_zone.hotplugged_size - size,
+                                user_addr: virtio_mem_region.as_ptr() as u64 + size,
+                            },
+                            false,
+                        ));
+                    }
+                    Ordering::Equal => {}
+                }
+
                 virtio_mem_zone
                     .resize_handler()
                     .work(size)
                     .map_err(Error::VirtioMemResizeFail)?;
+
+                // Update the hotplugged size
+                virtio_mem_zone.hotplugged_size = size;
             } else {
                 error!("Failed resizing virtio-mem region: No virtio-mem handler");
                 return Err(Error::MissingVirtioMemHandler);
             }
 
-            return Ok(());
+            return Ok(range);
         }
 
         error!("Failed resizing virtio-mem region: Unknown memory zone");
         Err(Error::UnknownMemoryZone)
     }
 
-    /// In case this function resulted in adding a new memory region to the
-    /// guest memory, the new region is returned to the caller. The virtio-mem
-    /// use case never adds a new region as the whole hotpluggable memory has
-    /// already been allocated at boot time.
-    pub fn resize(&mut self, desired_ram: u64) -> Result<Option<Arc<GuestRegionMmap>>, Error> {
+    pub fn resize_virtio_mem(&mut self, desired_ram: u64) -> Result<Option<(Range, bool)>, Error> {
+        if self.user_provided_zones {
+            error!(
+                "Not allowed to resize guest memory when backed with user \
+                defined memory zones."
+            );
+            return Err(Error::InvalidResizeWithMemoryZones);
+        }
+
+        let mut range: Option<(Range, bool)> = None;
+        if desired_ram >= self.boot_ram {
```

Comment:
Should this be > rather than >= ?

Reply:
No `>=` is the proper one. In case the user started with a 2G VM, then asked for 4G (which means we plugged 2G), and then later ask for 2G again, we need this to be valid so that it will be translated into: "Unplug 2G from virtio-mem".